flatflow/scheduler/internal/algorithm/regression.h

```diff
@@ -0,0 +1,109 @@
+// Copyright 2024 The FlatFlow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FLATFLOW_SCHEDULER_INTERNAL_ALGORITHM_REGRESSION_H_
+#define FLATFLOW_SCHEDULER_INTERNAL_ALGORITHM_REGRESSION_H_
+
+#include <omp.h>
+
+#include <cmath>
+#include <execution>
+#include <numeric>
+#include <vector>
+
+#include "absl/log/check.h"
+
+namespace flatflow {
+namespace scheduler {
+namespace internal {
+namespace algorithm {
+
+// PassiveAggressiveRegressor<>
+//
+// A`flatflow::scheduler::internal::algorithm::PassiveAggressiveRegressor` is
+// pure cpp version of sklearn.linear_model.PassiveAggressiveRegressor.
+//
+// coef_ is initialized in to 1.0. And bias term is initialized to 0.0.
+// This initialization method is based on flatflow static scheduler.
+// This Regressor is currently supporting basic PassiveAggressiveRegressor.
+// C_ and max_iter are each fixed to 1.0 and 1000.
+// Epsilon is the threshold for the update. Original PassiveAggressiveRegressor
+// assigns epsilon value to 0.1. Smaller the epsilon is, loss will be updated more.
+// See _sgd_fast.pyx.tp for detail implementation.
+//
+// References:
+// https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.PassiveAggressiveRegressor.html#sklearn.linear_model.PassiveAggressiveRegressor
+// https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/linear_model/_sgd_fast.pyx.tp
+template <std::size_t degree>
+class PassiveAggressiveRegressor {
+ public:
+  inline explicit PassiveAggressiveRegressor(const double &epsilon)
+      : epsilon_(epsilon), coef_(degree + 1, 1.0), power_(degree + 1, 0.0) {
+    coef_[0] = 0.0;
+  }
+
+  // PassiveAggressiveRegressor::fit()
+  //
+  // Fits the model to the given vector data X and y.
+  // This implementation is based on Epsilon-Insensitive Passive Aggressive Regressor.
+  // This method is implementation of partialfit method in order to use in
+  // online learning.
+  inline void fit(const std::vector<double> &X, const std::vector<double> &y) {
+    CHECK(X.size() == y.size());
+    for (std::size_t iter = 0; iter < kMaxIter_; ++iter) {
+      for (std::size_t dataIdx = 0; dataIdx < X.size(); ++dataIdx) {
+        const auto prediction = predict(X[dataIdx]);
+        const auto loss =
+            std::max(0.0, std::abs(y[dataIdx] - prediction) - epsilon_);
```

Comment:
Does it work doing ``std::max()`` with zero after ``std::abs()``?

Reply:
Yes. Loss is trying to get max value between an |y- prediction| and epsilon.
If epsilon is bigger than the |y-prediction| , loss becomes 0.0